lib/mix/lib/mix/scm/git.ex

```diff
@@ -250,24 +249,18 @@ defmodule Mix.SCM.Git do
   end
 
   defp validate_depth(opts) do
-    case Keyword.take(opts, [:depth, :ref]) do
-      [_, _] ->
-        Mix.raise(
-          "Cannot specify :depth and :ref at the same time. " <>
-            "Error on Git dependency: #{redact_uri(opts[:git])}"
-        )
+    case Keyword.take(opts, [:depth]) do
```

Comment:
We probably still want to validate here that the ref is a full ref and raise otherwise.

Reply:
I drafted something in that direction but gave up when reading the Git manual about future SHA256 hashes.

We never validated that a `:ref` is a hex object name. Do you think we should start being more strict now?

I did consider giving a more specific warning if the operation fails ("if you use :ref and :depth, use a full :ref to be able to resolve it"), in addition to the docs note, but didn't try implementing it.

---

Also, for the most part, we can use `:ref`, `:branch` and `:tag` kind of interchangeably -- In 1.17.2, I was able to use "ref" and `:depth` together by specifying `tag: SOME_FULL_SHA1, depth: 1` in mix.exs.

Similarly, one can write a tag name in `:ref` and it is going to work, with or without `:depth`. So I thought enforcing a 40-char `:ref` to be over-specific and limiting.

Reply:
Don't you find it brittle?

```ex
  defp validate_depth(opts) do
    case Keyword.take(opts, [:depth, :ref]) do
      [_, ref] ->
        if !Strings.length(ref) == 40 do
          Mix.raise(
            "When :depth is used with :ref, a full 40-character commit hash is required. " <>
              "Error on Git dependency: #{redact_uri(opts[:git])}"
          )
        end
      # ...
```

I don't feel confident being so assertive about what a valid `ref` is. Sure we could define it more strictly than Git, but that has the potential to break people's project in the wild.

What if we instead gave a warning if the `git fetch` command failed?

Reply:
I don't think we will break people's code in the wild because no one is using the depth+ref option. It was released in 1.17.0+1.17.1 but it didn't work. 1.17.2 straight raises. If we are being too restrict, someone can open up a bug report. We can also do `byte_size(ref) >= 40`, so we support larger refs in the future.

Reply:
You are right.

Updating the commit with this change:

```diff
diff --git lib/mix/lib/mix/scm/git.ex lib/mix/lib/mix/scm/git.ex
index 48efa92a3..442c241f5 100644
--- lib/mix/lib/mix/scm/git.ex
+++ lib/mix/lib/mix/scm/git.ex
@@ -248,12 +248,23 @@ defp validate_refspec(opts) do
     end
   end
 
+  @sha1_size 40
+
   defp validate_depth(opts) do
     case Keyword.take(opts, [:depth]) do
       [] ->
         opts
 
       [{:depth, depth}] when is_integer(depth) and depth > 0 ->
+        ref = opts[:ref]
+
+        if ref && byte_size(ref) < @sha1_size do
+          Mix.raise(
+            "When :depth is used with :ref, a full commit hash is required. " <>
+              "Error on Git dependency: #{redact_uri(opts[:git])}"
+          )
+        end
+
         opts
 
       invalid_depth ->
```